reco_utils/dataset/pandas_df_utils.py

```diff
@@ -74,3 +75,99 @@ def filter_by(df, filter_by_df, filter_by_cols):
             filter_by_df.set_index(filter_by_cols).index
         )
     ]
+
+
+def libffm_converter(df, col_rating=DEFAULT_RATING_COL, filepath=None):
+    """Converts an input Dataframe (df) to another Dataframe (df) in libffm format. A text file of the converted
+    Dataframe is optionally generated.
+
+    Note:
+        The input dataframe is expected to represent the feature data in the following schema
+        |field-1|field-2|...|field-n|rating|
+        |feature-1-1|feature-2-1|...|feature-n-1|1|
+        |feature-1-2|feature-2-2|...|feature-n-2|0|
+        ...
+        |feature-1-i|feature-2-j|...|feature-n-k|0|
+        Where
+        1. each "field-*" is the column name of the dataframe (column of lable/rating is excluded), and
+        2. "feature-*-*" can be either a string or a numerical value, representing the categorical variable or
+        actual numerical variable of the feature value in the field, respectively.
+        3. If there are ordinal variables represented in int types, users should make sure these columns
+        are properly converted to string type.
+
+        The above data will be converted to the libffm format by following the convention as explained in
+        https://www.csie.ntu.edu.tw/~r01922136/slides/ffm.pdf
+
+        i.e., <field_index>:<field_feature_index>:1 or <field_index>:<field_index>:<field_feature_value>, depending on
+        the data type of the features in the original dataframe.
+
+    Examples:
+        >>> import pandas as pd
+        >>> df_feature = pd.DataFrame({
+                'rating': [1, 0, 0, 1, 1],
+                'field1': ['xxx1', 'xxx2', 'xxx4', 'xxx4', 'xxx4'],
+                'field2': [3, 4, 5, 6, 7],
+                'field3': [1.0, 2.0, 3.0, 4.0, 5.0],
+                'field4': ['1', '2', '3', '4', '5']
+            })
+
+        >>> df_out = libffm_converter(df_feature, col_rating='rating')
+        >>> df_out
+            rating field1 field2   field3 field4
+        0       1  1:1:1  2:2:3  3:3:1.0  4:4:1
```

Comment:
I think the feature indexes should be distinct. See slide 13 in https://www.csie.ntu.edu.tw/~r01922136/slides/ffm.pdf
You reuse however feature 2 and feature 3 for field1, field2, field3.